File: nim/src/benchmarks/matmul1t.nim
import std/[math, strutils]
import ../benchmark
import ../helper

type
  Matmul1T* = ref object of Benchmark
    n: int64
    resultVal: uint32

proc newMatmul1T(): Benchmark =
  Matmul1T()

method name(self: Matmul1T): string = "Matmul1T"

method prepare(self: Matmul1T) =
  self.n = self.config_val("n")
  self.resultVal = 0

proc matgen(n: int): seq[seq[float64]] {.inline.} =
  let tmp = 1.0 / float64(n * n)

  result = newSeq[seq[float64]](n)
  for i in 0..<n:
    result[i] = newSeq[float64](n)

  {.push checks: off.}
  for i in 0..<n:
    for j in 0..<n:
      result[i][j] = tmp * float64(i - j) * float64(i + j)
  {.pop.}

proc matmul_single_thread(a, b: seq[seq[float64]]): seq[seq[float64]] {.inline.} =
  let n = a.len

  var bT = newSeq[seq[float64]](n)
  for j in 0..<n:
    bT[j] = newSeq[float64](n)

  {.push checks: off.}
  for i in 0..<n:
    for j in 0..<n:
      bT[j][i] = b[i][j]
  {.pop.}

  result = newSeq[seq[float64]](n)
  for i in 0..<n:
    result[i] = newSeq[float64](n)

  {.push checks: off.}
  for i in 0..<n:
    let ai = a[i]
    for j in 0..<n:
      let bj = bT[j]
      var sum = 0.0

      for k in 0..<n:
        sum += ai[k] * bj[k]

      result[i][j] = sum
  {.pop.}

method run(self: Matmul1T, iteration_id: int) =
  let nInt = int(self.n)
  let a = matgen(nInt)
  let b = matgen(nInt)
  let c = matmul_single_thread(a, b)

  let centerIdx = nInt shr 1
  self.resultVal = self.resultVal + checksumF64(c[centerIdx][centerIdx])

method checksum(self: Matmul1T): uint32 =
  self.resultVal

registerBenchmark("Matmul1T", newMatmul1T)

File: rust/src/benchmarks/matmul1t.rs
use super::super::{helper, Benchmark};
use crate::config_i64;

pub struct Matmul1T {
    n: i64,
    result_val: u32,
}

impl Matmul1T {
    pub fn new() -> Self {
        let n = config_i64("Matmul1T", "n");

        Self { n, result_val: 0 }
    }

    fn matmul(&self, a: &[Vec<f64>], b: &[Vec<f64>]) -> Vec<Vec<f64>> {
        let m = a.len();
        let n = a[0].len();
        let p = b[0].len();

        let mut b2 = vec![vec![0.0; n]; p];
        for i in 0..n {
            for j in 0..p {
                b2[j][i] = b[i][j];
            }
        }

        let mut c = vec![vec![0.0; p]; m];
        for i in 0..m {
            let ai = &a[i];
            let ci = &mut c[i];

            for j in 0..p {
                let mut s = 0.0;
                let b2j = &b2[j];

                for k in 0..n {
                    s += ai[k] * b2j[k];
                }

                ci[j] = s;
            }
        }

        c
    }

    fn matgen(&self, n: usize) -> Vec<Vec<f64>> {
        let tmp = 1.0 / (n * n) as f64;
        let mut a = vec![vec![0.0; n]; n];

        for i in 0..n {
            for j in 0..n {
                a[i][j] = tmp * (i as f64 - j as f64) * (i as f64 + j as f64);
            }
        }

        a
    }
}

impl Benchmark for Matmul1T {
    fn name(&self) -> String {
        "Matmul1T".to_string()
    }

    fn run(&mut self, _iteration_id: i64) {
        let n = self.n as usize;
        let a = self.matgen(n);
        let b = self.matgen(n);
        let c = self.matmul(&a, &b);

        let center_value = c[n >> 1][n >> 1];
        self.result_val = self
            .result_val
            .wrapping_add(helper::checksum_f64(center_value));
    }

    fn checksum(&self) -> u32 {
        self.result_val
    }
}
